### ISSUE TITLE
CLI: `sourcesContent` field in `.map` file should always contain source code

### ISSUE THREAD

**Issue by mems** [jashkenas/coffeescript#4342]
Actually we've got :
1. `sourceMap`: enable source map, create additional .map contains map between generated file and the original source file
2. `inlineMap` (previously `inline`): enable source map, inline source map in generated file with original source file content

We need a third option to activate separately source file content inlining (+ add clarity): 
1. `sourceMap`: enable source map, create additional `.map` file contains source map data
2. `inlineSourceMap`: enable source map, inline source map data in the generated file
3. `inlineSourceMapSource`: inline in source map data the source file content (add `sourcesContent` field)

Related #4223

**Comment by danielbayley** [comment]
Maybe it could instead match Babel's `sourceMap` option (`[ true, false, 'inline', 'both' ]`?

**Comment by GeoffreyBooth** [comment]
There are lots of downstream tools that rely on the source map options that exist today, so we shouldn’t change the existing API if we don’t need to. But that doesn’t mean we can’t add a new option.

**Comment by GeoffreyBooth** [comment]
I think I misread this before. So currently, specifying both options will produce both source maps:
```bash
coffee --compile --output test.js --map --inline-map --no-header --bare test.coffee
```
results in both `test.js`:
```js
console.log('woohoo!');

//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoidGVzdC5qcyIsInNvdXJjZVJvb3QiOiIiLCJzb3VyY2VzIjpbInRlc3QuY29mZmVlIl0sIm5hbWVzIjpbXSwibWFwcGluZ3MiOiJBQUFBLE9BQU8sQ0FBQyxHQUFSLENBQVksU0FBWiIsInNvdXJjZXNDb250ZW50IjpbImNvbnNvbGUubG9nICd3b29ob28hJ1xuIl19
//# sourceURL=~/Sites/coffeescript/test.coffee
//# sourceMappingURL=test.js.map
```
and `test.js.map`.
```json
{
  "version": 3,
  "file": "test.js",
  "sourceRoot": "",
  "sources": [
    "test.coffee"
  ],
  "names": [],
  "mappings": "AAAA,OAAO,CAAC,GAAR,CAAY,SAAZ",
  "sourcesContent": [
    "console.log 'woohoo!'\n"
  ]
}
```

So the ask isn’t that there be an option to generate both, since that’s already possible. That was my initial misreading.

Looking at it again, it sounds like what you’re asking for is that the `sourcesContent` field in the inline source map is filled in with the CoffeeScript code content. But as of 841b3cd2ad7e1eea8115a7d1e5ceafdf01bb9259, that’s already happening, as you can see above; and that commit was 7 months before this issue was posted. So what’s the problem with the current output? The fact that enabling both options leaves two comments at the bottom?

**Comment by mems** [comment]
As I can read in [sourcemap.litcoffee](https://github.com/jashkenas/coffeescript/blob/master/src/sourcemap.litcoffee), `sourcesContent` field is only added in source map file, if you add the option `inlineMap`:
```
v3.sourcesContent = [code] if options.inlineMap
```

Than means, if you need the source map file, but not the inlined source map in the generated JS file (`//# sourceMappingURL=data:application/json;base64,... //# sourceURL=...` comments) you can't.
You need an external tool to remove this comment (extra bytes) to left only the comment part `//# sourceMappingURL=test.js.map`

**Comment by GeoffreyBooth** [comment]
It seems like the fact that `sourcesContent` is empty in `.map` files, unless inline maps are also generated, is a bug. `sourcesContent` should always have the source, whichever option you choose. That’s Babel’s behavior:

```json
 ✦ ./node_modules/babel-cli/bin/babel.js --presets env --source-maps \
--out-file test.babel.js test.js
 ✦ node -e "console.log(JSON.stringify(JSON.parse(require('fs').readFileSync(
'./test.babel.js.map', 'utf-8')), null, 2))"
{
  "version": 3,
  "sources": [
    "test.js"
  ],
  "names": [],
  "mappings": ";;AAAA;;;;;;AACA,QAAQ,GAAR,CAAY,eAAK,GAAjB",
  "file": "test.babel.js",
  "sourcesContent": [
    "import path from \"path\"\nconsole.log(path.sep)\n"
  ]
}
```

**Comment by mems** [comment]
Yes, that is the right behavior

**Comment by GeoffreyBooth** [comment]
Fixed via #4698.
